velox/type/tests/TimestampTest.cpp

```diff
@@ -299,37 +299,133 @@ TEST(TimestampTest, outOfRange) {
+    oss << '.' << std::setfill('0') << std::setw(width) << value;
+  }
+
+  return oss.str();
+}
+
 TEST(TimestampTest, epochToUtc) {
-  std::tm tm;
-  ASSERT_FALSE(epochToUtc(-(1ll << 60), tm));
-  ASSERT_FALSE(epochToUtc(1ll << 60, tm));
+  std::tm tm{};
+  ASSERT_FALSE(Timestamp::epochToUtc(-(1ll << 60), tm));
+  ASSERT_FALSE(Timestamp::epochToUtc(1ll << 60, tm));
 }
 
 TEST(TimestampTest, randomEpochToUtc) {
   uint64_t seed = 42;
-  // seed = std::random_device{}();
   std::default_random_engine gen(seed);
   std::uniform_int_distribution<time_t> dist(
       std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
-  std::tm actual, expected;
-  for (int i = 0; i < 10000; ++i) {
+  std::tm actual{};
+  std::tm expected{};
+  for (int i = 0; i < 10'000; ++i) {
     auto epoch = dist(gen);
     SCOPED_TRACE(fmt::format("epoch={}", epoch));
     if (gmtime_r(&epoch, &expected)) {
-      ASSERT_TRUE(epochToUtc(epoch, actual));
-      ASSERT_EQ(expected.tm_year, actual.tm_year);
-      ASSERT_EQ(expected.tm_yday, actual.tm_yday);
-      ASSERT_EQ(expected.tm_mon, actual.tm_mon);
-      ASSERT_EQ(expected.tm_mday, actual.tm_mday);
-      ASSERT_EQ(expected.tm_wday, actual.tm_wday);
-      ASSERT_EQ(expected.tm_hour, actual.tm_hour);
-      ASSERT_EQ(expected.tm_min, actual.tm_min);
-      ASSERT_EQ(expected.tm_sec, actual.tm_sec);
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
     } else {
-      ASSERT_FALSE(epochToUtc(epoch, actual));
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
     }
   }
 }
 
+void checkTimeString(
+    const std::string& format,
+    const TimestampToStringOptions& options,
+    bool isGenNanos = true) {
+  uint64_t seed = 42;
+  std::default_random_engine gen(seed);
+  std::default_random_engine nanosGen(seed);
+
+  std::uniform_int_distribution<time_t> dist(
+      std::numeric_limits<time_t>::min(), std::numeric_limits<time_t>::max());
+  std::uniform_int_distribution<int> nanosDist(0, Timestamp::kMaxNanos);
+
+  std::tm actual{};
+  std::tm expected{};
+
+  for (int i = 0; i < 10'000; ++i) {
+    auto epoch = dist(gen);
+    auto nanos = isGenNanos ? nanosDist(nanosGen) : 0;
+    SCOPED_TRACE(fmt::format("epoch={}", epoch));
+    if (gmtime_r(&epoch, &expected)) {
+      ASSERT_TRUE(Timestamp::epochToUtc(epoch, actual));
+      checkTm(actual, expected);
+
+      auto actualMillisStr = Timestamp::tmToString(actual, nanos, options);
+      auto expectedMillisStr = tmToString(expected, nanos, format, options);
+      ASSERT_EQ(expectedMillisStr, actualMillisStr);
+
+    } else {
+      ASSERT_FALSE(Timestamp::epochToUtc(epoch, actual));
+    }
+  }
+}
+
+TEST(TimestampTest, tmToStringDateOnly) {
+  TimestampToStringOptions options;
+  options.mode = TimestampToStringOptions::Mode::kDateOnly;
+  checkTimeString("%F", options, false);
+
+  // With nanoseconds, the output date string will not contain nanoseconds
```

Comment:
We only format date, right? Why does it matter whether there are nanoseconds present or not?

Is there any reason to have isGenNanos parameter in checkTimeString ? Why not always include nanoseconds?

nit: checkTimeString -> testTmToString

Reply:
> We only format date, right? Why does it matter whether there are nanoseconds present or not?

Just used to verify that adding nanoseconds will not affect the result of date. Let me modify `checkTimeString` so that it generates nanoseconds every time